pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,282 @@
+type PolicyRuleWithSubjects struct {
+	// `subjects` is the list of normal user, serviceaccount, or group that this rule cares about.
+	// +optional
+	Subjects []Subject
+	// `rule` is the target verb, resource or the subresource the rule cares about. APIGroups, Resources, etc.
+	// Required.
+	Rule PolicyRule
+}
+
+// FlowSchemaStatus represents the current state of a flow-schema.
+type FlowSchemaStatus struct {
+	// Current state of flow-schema.
+	Conditions []FlowSchemaCondition
+}
+
+// FlowSchemaCondition describes conditions for a flow-schema.
+type FlowSchemaCondition struct {
+	// `type` is the type of the condition.
+	Type FlowSchemaConditionType
+	// `status` is the status of the condition.
+	// Can be True, False, Unknown.
+	Status corev1.ConditionStatus
+	// `lastTransitionTime` is the last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	// `reason` is a unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string
+	// `message` is a human-readable message indicating details about last transition.
+	Message string
+}
+
+// FlowSchemaConditionType is a valid value for FlowSchemaStatusCondition.Type
+type FlowSchemaConditionType string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// `spec` is the specification of the desired behavior of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec
+	// `status` is the current status of a "request-priority".
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+type PriorityLevelConfigurationSpec struct {
+	// `globalDefault` specifies whether this priority level should be considered as the default priority for requests
+	// that do not have any priority level. Only one PriorityClass should be marked as `globalDefault`.
+	// +optional
+	GlobalDefault *bool
```

Comment:
We do not need these to be pointers.  I realize that the end of https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required _requests_ the use of pointers even when the zero value is a good default.  This request is quite often violated in kubernetes already.  Also, the vision is dubious.  Imagine an evolutionary scenario where producers and consumers agree that nil means the same thing as zero.  Then suddenly some parts of the system start thinking differently.  I doubt this will go well.  Evolution needs a different vision than that.  So I am really not convinced that this is a good request.

Reply:
Note also that the request that "pointers always be used with optional fields that do not have a built-in `nil` value" applies to `string`-valued fields too.